Fix interpolation of convnext image processor

chandrusuresh opened the pull request:
# What does this PR do?
Image processors do not have the same interpolation method used in the actual models (reference to actual models: timm). This PR fixes:
- the interpolation method for ConvNext
- updates the conversion script to use timm models instead of downloading checkpoints from url. This was necessary to overcome the AssertionError thrown from comparing the logits to the hard-coded logits from the checkpoint url and fixes the bug in #37461 
- Convnext model (in timm) seems to use `IMAGENET_DEFAULT_MEAN/STD` to normalize image. But the default value set in `ConvNextImageProcessor` and `ConvNextImageProcessorFast` are `IMAGENET_STANDARD_MEAN/STD`. This has been fixed in this PR.

For the 2nd point above, the following timm models were tested to pass all assertions in the script. The following table summarizes the equivalence between the timm models and the checkpoint urls in the original script.

### Some noteworthy changes
(!) Default interpolator was updated on `ConvNextImageProcessorFast`. But `timm_pixel_values` do not match and raise an AssertionError. This is outside the scope of this PR.

| Checkpoint URL      | TIMM model name |
| ----------- | ----------- |
| [convnext_tiny_1k_224_ema.pth](https://dl.fbaipublicfiles.com/convnext/convnext_tiny_1k_224_ema.pth) | [convnext_tiny.fb_in1k](https://huggingface.co/timm/convnext_tiny.fb_in1k) |
| [convnext_small_1k_224_ema.pth](https://dl.fbaipublicfiles.com/convnext/convnext_small_1k_224_ema.pth) | [convnext_small.fb_in1k](https://huggingface.co/timm/convnext_small.fb_in1k) |
| [convnext_base_1k_224_ema.pth](https://dl.fbaipublicfiles.com/convnext/convnext_base_1k_224_ema.pth) | [convnext_base.fb_in1k](https://huggingface.co/timm/convnext_base.fb_in1k) |
| [convnext_base_1k_384.pth](https://dl.fbaipublicfiles.com/convnext/convnext_base_1k_384.pth) | Not available |
| [convnext_large_1k_224_ema.pth](https://dl.fbaipublicfiles.com/convnext/convnext_large_1k_224_ema.pth) | [convnext_large.fb_in1k](https://huggingface.co/timm/convnext_large.fb_in1k) |
| [convnext_large_1k_384.pth](https://dl.fbaipublicfiles.com/convnext/convnext_large_1k_384.pth) | Not available |
| [convnext_base_22k_224.pth](https://dl.fbaipublicfiles.com/convnext/convnext_base_22k_224.pth) | [convnext_base.fb_in22k](https://huggingface.co/timm/convnext_base.fb_in22k) |
| [convnext_large_22k_224.pth](https://dl.fbaipublicfiles.com/convnext/convnext_large_22k_224.pth) | [convnext_large.fb_in22k](https://huggingface.co/timm/convnext_large.fb_in22k) |
| [convnext_xlarge_22k_224.pth](https://dl.fbaipublicfiles.com/convnext/convnext_xlarge_22k_224.pth) | [convnext_xlarge.fb_in22k](https://huggingface.co/timm/convnext_xlarge.fb_in22k) |
| [convnext_base_22k_1k_224.pth](https://dl.fbaipublicfiles.com/convnext/convnext_base_22k_1k_224.pth) | [convnext_base.fb_in22k_ft_in1k](https://huggingface.co/timm/convnext_base.fb_in22k_ft_in1k) |
| [convnext_base_22k_1k_384.pth](https://dl.fbaipublicfiles.com/convnext/convnext_base_22k_1k_384.pth) | [convnext_base.fb_in22k_ft_in1k_384](https://huggingface.co/timm/convnext_base.fb_in22k_ft_in1k_384) |
| [convnext_large_22k_1k_224.pth](https://dl.fbaipublicfiles.com/convnext/convnext_large_22k_1k_224.pth) | [convnext_large.fb_in22k_ft_in1k](https://huggingface.co/timm/convnext_large.fb_in22k_ft_in1k) |
| [convnext_large_22k_1k_384.pth](https://dl.fbaipublicfiles.com/convnext/convnext_large_22k_1k_384.pth) | [convnext_large.fb_in22k_ft_in1k_384](https://huggingface.co/timm/convnext_large.fb_in22k_ft_in1k_384) |
| [convnext_xlarge_22k_1k_224_ema.pth](https://dl.fbaipublicfiles.com/convnext/convnext_xlarge_22k_1k_224_ema.pth) | [convnext_xlarge.fb_in22k_ft_in1k](https://huggingface.co/timm/convnext_xlarge.fb_in22k_ft_in1k) |
| [convnext_xlarge_22k_1k_384_ema.pth](https://dl.fbaipublicfiles.com/convnext/convnext_xlarge_22k_1k_384_ema.pth) | [convnext_xlarge.fb_in22k_ft_in1k_384](https://huggingface.co/timm/convnext_xlarge.fb_in22k_ft_in1k_384) |

The following were not tested:
- `model.save_pretrained` and `image_processor.save_pretrained`
- `model.push_to_hub`

Fixes #28180 , #37461

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?
@NielsRogge @amyeroberts

chandrusuresh committed:
Bicubic interpolation for convnext

chandrusuresh committed:
Merge branch 'huggingface:main' into chandra/fix_interpolation_image_processors

chandrusuresh committed:
Fix local changes

chandrusuresh committed:
Merge branch 'chandra/fix_interpolation_image_processors' of https://github.com/chandrusuresh/transformers into chandra/fix_interpolation_image_processors

chandrusuresh committed:
fix ruff

chandrusuresh committed:
fix formatting

chandrusuresh committed:
Fix resampling in fast processor

chandrusuresh committed:
Compare with timm directly

chandrusuresh committed:
Merge branch 'main' into chandra/fix_interpolation_image_processors

chandrusuresh committed:
fix ruff formatting

chandrusuresh committed:
Merge branch 'chandra/fix_interpolation_image_processors' of https://github.com/chandrusuresh/transformers into chandra/fix_interpolation_image_processors

chandrusuresh committed:
Fix image mean/std to imagenet default values

chandrusuresh committed:
set default values with minimal change

chandrusuresh committed:
ruff formatting

chandrusuresh committed:
Merge branch 'main' into chandra/fix_interpolation_image_processors

chandrusuresh committed:
Merge branch 'main' into chandra/fix_interpolation_image_processors

chandrusuresh committed:
Merge branch 'main' into chandra/fix_interpolation_image_processors

qubvel reviewed (commented):
Hey @chandrusuresh! Thanks a lot for opening a PR

I agree regarding bicubic interpolation and DEAFULT mean/std changes, however not sure regarding changes in the conversion script. Originally, checkpoints were converted from facebookresearch URLs, so it's better to keep them.

chandrusuresh commented:
> Hey @chandrusuresh! Thanks a lot for opening a PR
> 
> I agree regarding bicubic interpolation and DEAFULT mean/std changes, however not sure regarding changes in the conversion script. Originally, checkpoints were converted from facebookresearch URLs, so it's better to keep them.

Thanks for the response. The issue is not really the checkpoints, but more the check with the logits. The hard-coded values do not match with the output of the image processor. 
There are other image processors that use the timm models as a baseline for comparing the image processors, so I decided to do the same.
If we keep the checkpoints as is, we need to get the logits somehow. any thoughts how we may do that?

qubvel commented:
I suppose they can be obtained running the original implementation

https://github.com/facebookresearch/ConvNeXt/tree/main

chandrusuresh committed:
revert to using checkpoints, but do not use hard-coded logits

chandrusuresh committed:
fix ruff formatting

chandrusuresh committed:
fix ruff formatting again

chandrusuresh committed:
Merge branch 'main' into chandra/fix_interpolation_image_processors

chandrusuresh commented:
> I suppose they can be obtained running the original implementation
> 
> https://github.com/facebookresearch/ConvNeXt/tree/main

Thanks for the feedback. I've reverted the change back to using checkpoint_url. I replaced the hard-coded logits (which only have 3 elements) with those computed from timm models. I've verified that all the checkpoints in the above table pass the asserts.

chandrusuresh committed:
Merge branch 'main' into chandra/fix_interpolation_image_processors

chandrusuresh commented:
@qubvel  I'd appreciate a review here. I've made the changes as requested.

chandrusuresh committed:
Merge branch 'main' into chandra/fix_interpolation_image_processors

chandrusuresh commented:
@ArthurZucker @Rocketknight1 @qubvel any chance you could review this? its been stale for more than a week now. I'd appreciate a quick review on this, pls?

chandrusuresh committed:
Merge branch 'main' into chandra/fix_interpolation_image_processors

NielsRogge commented:
Hi @chandrusuresh thanks for this PR and apologies for getting no response here. The PR looks good to me. Could you rebase on main by doing:
```bash
git remote add upstream https://github.com/huggingface/transformers.git
git fetch upstream
git merge upstream/main
```
?

chandrusuresh committed:
Merge remote-tracking branch 'upstream/main' into chandra/fix_interpolation_image_processors

chandrusuresh committed:
Merge branch 'chandra/fix_interpolation_image_processors' of https://github.com/chandrusuresh/transformers into chandra/fix_interpolation_image_processors

chandrusuresh commented:
@NielsRogge Thanks for the approval. Merged to upstream and ready for merge.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_37460). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

chandrusuresh committed:
Merge branch 'main' into chandra/fix_interpolation_image_processors

chandrusuresh commented:
@qubvel @Rocketknight1 @NielsRogge  Can I get an approval on this PR, please?

chandrusuresh committed:
Merge branch 'main' into chandra/fix_interpolation_image_processors

chandrusuresh committed:
Merge branch 'main' into chandra/fix_interpolation_image_processors

NielsRogge commented:
Hi @qubvel could you review this one?

chandrusuresh committed:
Merge branch 'main' into chandra/fix_interpolation_image_processors

chandrusuresh committed:
Merge branch 'main' into chandra/fix_interpolation_image_processors

qubvel commented on src/transformers/models/convnext/convert_convnext_to_pytorch.py:
```diff
@@ -223,10 +309,9 @@ def convert_convnext_checkpoint(checkpoint_url, pytorch_dump_folder_path):
     parser.add_argument(
         "--checkpoint_url",
-        default="https://dl.fbaipublicfiles.com/convnext/convnext_tiny_1k_224_ema.pth",
+        default="http://dl.fbaipublicfiles.com/convnext/convnext_tiny_1k_224_ema.pth",
```
Can you please explain what is the purpose of this change?

chandrusuresh replied:
I had some trouble testing it. I just reverted this change in the latest commit.
I reverted the change to use timm models based on your previous feedback.
I can make a new PR for that next after this gets merged. Thanks!

qubvel reviewed (commented):
Thanks for working on this! Along with adding timm models conversion, I think we should also keep the conversion of the original (Facebook) models. Alternatively this can be split into two separate PRs, one for updating the default mean/std (breaking change), and another one for updating the conversion script.

chandrusuresh committed:
fix https

chandrusuresh committed:
Merge branch 'main' into chandra/fix_interpolation_image_processors

chandrusuresh committed:
Merge branch 'main' into chandra/fix_interpolation_image_processors

ArthurZucker commented:
@chandrusuresh do you need another round of reviewas?
